tsl/src/remote/connection.c

```diff
@@ -531,11 +535,7 @@ get_libpq_options()
 		libpq_options = PQconndefaults();
 	}
 
-	if (libpq_options == NULL)
-	{
-		/* probably OOM */
-		elog(ERROR, "could not get default libpq options");
-	}
+	Ensure(libpq_options, "out of memory");
```

Comment:
This might be the most common error, as an improvement I suggest to make a new macro like `Ensure_oom` (or something like that) using `ERRCODE_OUT_OF_MEMORY` code underneath

Reply:
Good idea. Makes more sense than using `ERRCODE_INTERNAL_ERROR`.

Reply:
Why not just `ereport`? Ensure is for internal program errors, and OOM is an environment error.

Reply:
> Why not just `ereport`? Ensure is for internal program errors, and OOM is an environment error.

I think it is a matter of definitions. IMHO, `Ensure` is used when environmental issues can cause problems as well, such as bad catalog data.

No strong opinion on what to use here, but if using `ERRCODE_OUT_OF_MEMORY` it is simpler to replace with `ereport`.

Reply:
Not sure why this is an improvement over the previous error. The pointer was already checked for NULL and the new error message is more generic than the previous one.

Reply:
>  I think it is a matter of definitions. 

Yes, it would be good to work out some guidelines for this. Being able to distinguish program logic errors is important for fuzzing, like SQLSmith, or some kinds of stress testing. Things like `Assert` and `elog` are such errors, if you see them under any conditions, it's a bug. We could make SQLSmith more efficient by adding some handling for `ERRCODE_INTERNAL_ERROR` errors, and using `elog` vs `ereport` in a more principled manner. In the context of testing, catalog corruption is probably also an internal error? Not sure...

Reply:
> In the context of testing, catalog corruption is probably also an internal error? Not sure...

Oh, I think it is, but it can be caused by factors out of our control, so we should have checks for it and just not assume that it is always correct.

Reply:
Added a macro `TS_OOM_CHECK` and used that instead. It calls `ereport`.